src/wasm/wasm-type.cpp

```diff
@@ -2699,7 +2696,9 @@ bool RecGroupEquator::eq(HeapType a, HeapType b) const {
   }
   auto groupA = a.getRecGroup();
   auto groupB = b.getRecGroup();
-  return groupA == groupB || (groupA == newGroup && groupB == otherGroup);
+  bool selfRefA = groupA == newGroup;
+  bool selfRefB = groupB == otherGroup;
+  return (selfRefA && selfRefB) || (!selfRefA && !selfRefB && groupA == groupB);
```

Comment:
I've forgotten how this code works, sorry. Is `newGroup` a new group being built that is being compared to an existing group `otherGroup`? How can `groupA/B` not match those two, why would we be comparing types not from the groups at all, when comparing `newGroup/otherGroup`?

I couldn't find code comments explaining how the class `RecGroupEquator` is used - are they somewhere else than on the class itself?

Reply:
> I've forgotten how this code works, sorry. Is newGroup a new group being built that is being compared to an existing group otherGroup?

Yes, exactly.

> How can `groupA/B` not match those two, why would we be comparing types not from the groups at all, when comparing `newGroup/otherGroup`?

Because we might be comparing heap types that are used in `newGroup` and `otherGroup` (e.g. because structs defined in those groups contain reference to the heap types), but are defined in some arbitrary different previous groups. We're trying to compare the "shapes" of `newGroup` and `otherGroup`, and how they reference other groups is part of that "shape."

> I couldn't find code comments explaining how the class RecGroupEquator is used - are they somewhere else than on the class itself?

There are some additional comments on `RecGroupStructure` and `RecGroupStore`.

Reply:
Got it, thanks!